scripts/us_bea/states_gdp/import_data.py

```diff
@@ -0,0 +1,136 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Pulls data from the US Bureau of Economic Analysis (BEA) on quarterly GDP
+per US state. Saves output as a CSV file.
+
+    Typical usage:
+
+    python3 import_data.py
+"""
+from urllib.request import urlopen
+from absl import app
+import pandas as pd
+import io
+import zipfile
+import csv
+import re
+
+# Suppress annoying pandas DF copy warnings.
+pd.options.mode.chained_assignment = None # default='warn'
+
+class StateGDPDataLoader:
+    """Pulls per-state GDP data from the BEA.
+
+    Attributes:
+        df: DataFrame (DF) with the cleaned data.
+    """
+    _US_STATES = ['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
+                  'Colorado', 'Connecticut', 'Delaware', 'District of Columbia',
+                  'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana',
+                  'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
+                  'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
+                  'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
+                  'New Jersey', 'New Mexico', 'New York', 'North Carolina',
+                  'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
+                  'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
+                  'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
+                  'West Virginia', 'Wisconsin', 'Wyoming']
+    _ZIP_LINK = "https://apps.bea.gov/regional/zip/SQGDP.zip"
+    _STATE_QUARTERLY_GDP_FILE = "SQGDP1__ALL_AREAS_2005_2019.csv"
+
+    def __init__(self):
+        """Downloads the data, cleans it and stores it in instance data frame.
+
+        Returns
+        """
+        df = self.download_data()
+
+        # Filters out columns that are not US states (e.g. New England).
+        df = df[df['GeoName'].isin(self._US_STATES)]
+
+        # Gets columns that represent quarters, e.g. 2015:Q2, by matching
+        # against a regular expression.
+        all_quarters = [q for q in df.columns if re.match(r"....:Q.", q)]
+
+        # Convert table from wide to long format.
+        df = pd.melt(df, id_vars=['GeoFIPS', 'Unit'],
+                     value_vars=all_quarters,
+                     var_name='Quarter')
+
+        qtr_month_map = {
+            'Q1':'03',
+            'Q2':'06',
+            'Q3':'09',
+            'Q4':'12'
+        }
+
+        def date_to_obs_date(date):
+            """Converts date format, e.g., "2005:Q3" to "2005-09"."""
```

Comment:
Good practice to create proper docstrings:
http://google.github.io/styleguide/pyguide.html#383-functions-and-methods

Btw, even though this is only used here, it's helpful to pull helpers like these out. Then you can more easily add unittests.

Reply:
Docstring fixed. The only purpose of this function is to pass it to a pd.DataFrame.apply() call. It seems a bit overkill to separate it and test it on its own (if this fit in-line, I would have made it a lambda). Do you think it should still be pulled out?